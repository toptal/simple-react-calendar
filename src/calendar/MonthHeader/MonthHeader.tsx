import React, { FC, SyntheticEvent } from 'react'

import { IMonthHeaderButtonRenderProps, IMonthHeaderDayRenderProps, IMonthHeaderRenderProps } from '../../@types'
import { IGetDaysOfWeek } from '../helper'
import MonthHeaderButton from '../MonthHeaderButton'
import MonthHeaderDay from '../MonthHeaderDay'

export type Props = {
  activeMonth: Date
  blockClassName: string
  customRender?: IMonthHeaderRenderProps
  daysOfWeek: string[]
  getDaysOfWeek: (data: IGetDaysOfWeek) => string[]
  getMonthFormattedName: (day: Date) => string
  handleOnSwitchMonth: (event: SyntheticEvent<HTMLElement>) => void
  isMonthNextEnabled: boolean
  isMonthPrevEnabled: boolean
  renderMonthHeaderButton?: IMonthHeaderButtonRenderProps
  renderMonthHeaderDay?: IMonthHeaderDayRenderProps
  showDaysOfWeek: boolean
  showMonthName: boolean
  showMonthNext: boolean
  showMonthPrev: boolean
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

const MonthHeader: FC<Props> = (props) => {
  const {
    activeMonth,
    blockClassName,
    customRender,
    daysOfWeek,
    getDaysOfWeek,
    getMonthFormattedName,
    handleOnSwitchMonth,
    isMonthNextEnabled,
    isMonthPrevEnabled,
    renderMonthHeaderButton,
    renderMonthHeaderDay,
    showDaysOfWeek,
    showMonthName,
    showMonthNext,
    showMonthPrev,
    weekStartsOn,
  } = props

  if (customRender) {
    return customRender(props)
  }

  if (!showMonthPrev && !showMonthName && !showMonthNext && !showDaysOfWeek) return null

  return (
    <header className={`${blockClassName}-month_header`}>
      <div className={`${blockClassName}-month_header_wrapper`}>
        {showMonthPrev && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthPrevEnabled}
            onClick={handleOnSwitchMonth}
            type="prev"
          />
        )}
        {showMonthName && (
          <h3 className={`${blockClassName}-month_header_title`}>{getMonthFormattedName(activeMonth)}</h3>
        )}
        {showMonthNext && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthNextEnabled}
            onClick={handleOnSwitchMonth}
            type="next"
          />
        )}
      </div>
      <div className={`${blockClassName}-month_header_days`}>
        {showDaysOfWeek &&
          getDaysOfWeek({dayIndex: weekStartsOn, daysOfWeek}).map((day, index) => (
            <MonthHeaderDay
              blockClassName={blockClassName}
              customRender={renderMonthHeaderDay}
              day={day}
              isWeekend={index >= weekStartsOn + 5}
              key={day}
            />
          ))}
      </div>
    </header>
  )
}

MonthHeader.displayName = 'MonthHeader'

export default MonthHeader
