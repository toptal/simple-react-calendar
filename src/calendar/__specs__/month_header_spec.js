import React from 'react'

import ReactTestRenderer from 'react-test-renderer'
import MonthHeader from '../month_header'

import {shallow} from 'enzyme'

describe('MonthHeader', () => {
  let props, wrapper, tree
  const defaultProps = {
    activeMonth: new Date(2015, 7, 17),
    onMonthChange: () => {},
  }

  describe('#_switchMonth', () => {
    it('calls #onMonthChange', () => {
      props = Object.assign(defaultProps, {
        maxDate: new Date(2015, 7, 1),
        onMonthChange: jest.fn(),
      })
      wrapper = shallow(<MonthHeader {...props} />)
      wrapper.instance()._switchMonth(1)

      expect(props.onMonthChange).toBeCalledWith(new Date(2015, 8, 17))
    })
  })

  describe('#render', () => {
    it('renders <MonthHeader />', () => {
      tree = ReactTestRenderer.create(<MonthHeader {...defaultProps} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    context('when prop `blockClassName` is defined', () => {
      it('renders <MonthHeader />', () => {
        props = Object.assign(defaultProps, {blockClassName: 'cal'})
        tree = ReactTestRenderer.create(<MonthHeader {...defaultProps} />).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    context('when prop `minDate` is after prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        props = Object.assign(defaultProps, {minDate: new Date(2015, 8, 17)})
        tree = ReactTestRenderer.create(<MonthHeader {...defaultProps} />).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    context('when prop `maxDate` is before prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        props = Object.assign(defaultProps, {maxDate: new Date(2015, 6, 17)})
        tree = ReactTestRenderer.create(<MonthHeader {...defaultProps} />).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })
  })
})
