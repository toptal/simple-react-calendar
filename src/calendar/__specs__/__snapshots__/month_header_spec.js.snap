// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthHeader #render renders <MonthHeader /> 1`] = `
<div
  className="calendar-month_header"
>
  <button
    className="calendar-header_button is-prev"
    disabled={false}
    onClick={[Function]}
    title="Previous month"
    type="button"
  />
  <div
    className="calendar-month_header_title"
  >
    August 2015
  </div>
  <button
    className="calendar-header_button is-next is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Next month"
    type="button"
  />
</div>
`

exports[`MonthHeader #render when prop \`blockClassName\` is defined renders <MonthHeader /> 1`] = `
<div
  className="cal-month_header"
>
  <button
    className="cal-header_button is-prev"
    disabled={false}
    onClick={[Function]}
    title="Previous month"
    type="button"
  />
  <div
    className="cal-month_header_title"
  >
    August 2015
  </div>
  <button
    className="cal-header_button is-next is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Next month"
    type="button"
  />
</div>
`

exports[`MonthHeader #render when prop \`maxDate\` is before prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="cal-month_header"
>
  <button
    className="cal-header_button is-prev is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Previous month"
    type="button"
  />
  <div
    className="cal-month_header_title"
  >
    August 2015
  </div>
  <button
    className="cal-header_button is-next is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Next month"
    type="button"
  />
</div>
`

exports[`MonthHeader #render when prop \`minDate\` is after prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="cal-month_header"
>
  <button
    className="cal-header_button is-prev is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Previous month"
    type="button"
  />
  <div
    className="cal-month_header_title"
  >
    August 2015
  </div>
  <button
    className="cal-header_button is-next is-disabled"
    disabled={true}
    onClick={[Function]}
    title="Next month"
    type="button"
  />
</div>
`
